mscore/exportmidi.cpp

```diff
@@ -288,11 +288,18 @@ bool ExportMidi::write(const QString& name, bool midiExpandRepeats)
 
                         for (auto i = events.begin(); i != events.end(); ++i) {
                               const NPlayEvent& event = i->second;
-                              if (event.discard())
-                                    continue;
+
                               if (event.getOriginatingStaff() != staffIdx)
                                     continue;
 
+                              if (event.discard()) { // ignore noteoff but restrike noteon
```

Comment:
If I understand correctly, this gives us two note on messages but only one note off.  I gather this works for us, but is it potentially problematic for some synthesizers / MIDI players?  It seems that maybe ideally we'd actually move the note off later, to coincide with the longer note?

Reply:
We never send note off messages ever in MuseScore. Instead we send noteon with velocity 0. Here we add one of this event just before the noteon. The number of "noteon" and "noteoff" (or noteon with vel 0) is the same.  Is it clearer ?

Reply:
Put differently, in fixupMIDI we keep clean the first note on and the last note off for overlapping notes. For all the others in between (off or on) we mark them with the `discard` flag. 
Then while playing or saving these events, we check the forward flag, ignore the noteoff and for noteon we send noteoff+noteon.

I mark them with forward instead of deleting or adding events directly in the eventMap in order to have a reference when hightlighting notes.

Reply:
Ah, I see, so they are still matched - just the first note gets explicitly cut off just before the second strikes.  Maybe that's the cause of the glitch cadiz1 mentioned, but I'm inclined to accept that.

BTW, from time to time we get reports about MuseScore not seeing the note off or mote on velocity=0 messages from some keyboard or other, resulting in note input appearing not to work at all, or at best stacking notes onto a single chord.  I was looking for an issue about that and couldn't find one.  Did we ever get to the bottom of that?

Reply:
This is totally different but no I don't think the code for MIDI input changed for 2.2.